src/NServiceBus.Metrics.ServiceControl.AcceptanceTests/Tests/QueueLength/When_native_queue_length_is_reported.cs

```diff
@@ -0,0 +1,109 @@
+            .WithEndpoint<MonitoringSpy>()
+            .Done(c => c.QueueLengthReportReceived)
+            .Run(TimeSpan.FromSeconds(10))
+            .ConfigureAwait(false);
+    }
+ 
+    class Context : ScenarioContext
+    {
+        public bool QueueLengthReportReceived { get; set; }
+    }
+
+    class EndpointWithNativeQueueLengthSupport : EndpointConfigurationBuilder
+    {
+        public EndpointWithNativeQueueLengthSupport()
+        {
+            EndpointSetup<DefaultServer>(c =>
+            {
+                var monitoringSpyAddress = Conventions.EndpointNamingConvention(typeof(MonitoringSpy));
+
+                c.EnableMetrics().SendMetricDataToServiceControl(monitoringSpyAddress, TimeSpan.FromSeconds(1));
+                c.EnableFeature<NativeQueueLengthFeature>();
+            });
+        }
+
+        class NativeQueueLengthFeature : Feature
+        {
+            public NativeQueueLengthFeature()
+            {
+                DependsOn("NServiceBus.Metrics.ServiceControl.ReportingFeature");
+            }
+
+            protected override void Setup(FeatureConfigurationContext context)
+            {
+                context.RegisterStartupTask(b => new NativeQueueLengthReporter(b.Build<IReportNativeQueueLength>()));
+            }
+        }
+
+        class NativeQueueLengthReporter : FeatureStartupTask
+        {
+            IReportNativeQueueLength queueLengthReporter;
+
+            public NativeQueueLengthReporter(IReportNativeQueueLength queueLengthReporter)
+            {
+                this.queueLengthReporter = queueLengthReporter;
+            }
+
+            protected override Task OnStart(IMessageSession session)
+            {
+                queueLengthReporter.ReportQueueLength("queue", 10);
+
+                return Task.FromResult(0);
+            }
+
+            protected override Task OnStop(IMessageSession session)
+            {
+                return Task.FromResult(0);
+            }
+        }
+    }
+
+
+    class MonitoringSpy : EndpointConfigurationBuilder
+    {
+        public MonitoringSpy()
+        {
+            EndpointSetup<DefaultServer>(c =>
+            {
+                c.Pipeline.Register(typeof(RawMessageBehavior), "Transport message handler.");
+                c.UseSerialization<NewtonsoftSerializer>();
+                c.LimitMessageProcessingConcurrencyTo(1);
+            }).IncludeType<EndpointMetadataReport>();
+        }
+
+        public class RawMessageBehavior : Behavior<IIncomingPhysicalMessageContext>
+        {
+            public Context TestContext { get; set; }
+
+            public override Task Invoke(IIncomingPhysicalMessageContext context, Func<Task> next)
+            {
+                if (context.MessageHeaders.TryGetValue("NServiceBus.Metric.Type", out var metricType) && metricType == "QueueLength")
+                {
+                    TestContext.QueueLengthReportReceived = true;
```

Comment:
Why don't we assert for the values being sent in here? How are we sure that "queue" and 10 are being sent? Are we do it on the basis of the buffer underneath?

Reply:
Assetions added.